test/EstateRegistry.js

```diff
@@ -0,0 +1,352 @@
+      await estate.transferManyTokens(
+        estateId,
+        [1],
+        anotherUser,
+        sentByAnotherUser
+      )
+    })
+
+    it('operator can not transfer tokens out after deauth', async function() {
+      const estateId = await createUserEstateWithToken1()
+      await estate.setApprovalForAll(anotherUser, true, sentByUser)
+      await transferOut(estateId, 1, sentByAnotherUser)
+      await transferIn(estateId, 1)
+      await estate.setApprovalForAll(anotherUser, false, sentByUser)
+      await assertRevert(transferOut(estateId, 1, sentByAnotherUser))
+    })
+
+    it('operator can not transfer many tokens out after deauth', async function() {
+      const estateId = await createUserEstateWithToken1()
+      await estate.setApprovalForAll(anotherUser, true, sentByUser)
+      await transferOut(estateId, 1, sentByAnotherUser)
+      await transferIn(estateId, 1)
+      await estate.setApprovalForAll(anotherUser, false, sentByUser)
+      await assertRevert(
+        estate.transferManyTokens(estateId, [1], anotherUser, sentByAnotherUser)
+      )
+    })
+  })
+
+  describe('order of tokens is correctly accounted', async function() {
+    it('five in, middle out, one in, middle out', async function() {
+      const estateId = await createUserEstateWithNumberedTokens()
+      await assertNFTBalance(estate.address, 5)
+      await transferOut(estateId, 2)
+      await assertTokenIdAtIndex(estateId, 1, 5)
+      await transferIn(estateId, 2)
+      await assertTokenIdAtIndex(estateId, 4, 2)
+      await transferOut(estateId, 3)
+      await assertTokenIdAtIndex(estateId, 2, 2)
+    })
+
+    it('five in, empty, refill', async function() {
+      const estateId = await createUserEstateWithNumberedTokens()
+      await transferOut(estateId, 2)
+      await transferOut(estateId, 1)
+      await transferOut(estateId, 3)
+      await transferOut(estateId, 4)
+      await transferOut(estateId, 5)
+      await assertNFTBalance(estate.address, 0)
+      await transferIn(estateId, 2)
+      await transferIn(estateId, 1)
+      await transferIn(estateId, 3)
+      await transferIn(estateId, 4)
+      await transferIn(estateId, 5)
+      await assertNFTBalance(estate.address, 5)
+      await assertTokenIdAtIndex(estateId, 0, 2)
+      await assertTokenIdAtIndex(estateId, 1, 1)
+      await assertTokenIdAtIndex(estateId, 2, 3)
+      await assertTokenIdAtIndex(estateId, 3, 4)
+      await assertTokenIdAtIndex(estateId, 4, 5)
+    })
+  })
+
+  describe('tokens are correctly accounted through detection', async function() {
+    it('out, unsafe in, check last', async function() {
+      const estateId = await createUserEstateWithNumberedTokens()
+      await transferOut(estateId, 2)
+      await unsafeTransferIn(2)
+      await assertNFTBalance(estate.address, 5)
+      await assertNFTOwner(2, estate.address)
+      await assertEstateSize(estateId, 4)
+      await estate.ammendReceived(estateId, 2, sentByUser)
```

Comment:
It would be great to check for the emitted events.

Here are some references for how to do it:

https://github.com/decentraland/land/blob/a2f6353af5ff1fa90ccfb95f2621c0f39895b185/test/LANDRegistry.js#L16

https://github.com/decentraland/land/blob/a2f6353af5ff1fa90ccfb95f2621c0f39895b185/test/LANDRegistry.js#L342

Reply:
Agree 👍